### PR TITLE
Feature/new blueprints

### PR DESCRIPTION
**Pull request by wrp801**
New blueprints:
- Create SQL Runner Query
- Run SQL Runner Query
- Download Dashboard
- Download Look

### CODE REVIEW

File: looker_blueprints/download_dashboard.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    )
+    # generate SDK
+    look_sdk = helpers.get_sdk(base_url, client_id, client_secret)
+    
+    # download look and write to file
+    width = args.output_width
+    height = args.output_height
+    result = download_dashboard(look_sdk, dashboard_id, width, height, file_type)
+
+    with open(destination_file_path, 'wb+') as f:
```
Comment (blakeburch) [inline looker_blueprints/download_dashboard.py:111]: Needs to include the typical shipyard_utils logic for "create_folder_if_dne". Currently breaks if the folder doesn't exist.

File: looker_blueprints/helpers.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+base_folder_name = shipyard.logs.determine_base_artifact_folder(
+    'looker')
+artifact_subfolder_paths = shipyard.logs.determine_artifact_subfolders(
+    base_folder_name)
+shipyard.logs.create_artifacts_folders(artifact_subfolder_paths)
+
+
+class MyApiSettings(api_settings.ApiSettings):
+    def __init__(self, *args, **kw_args):
+        self.base_url = kw_args.pop('base_url')
```
Comment (blakeburch) [inline looker_blueprints/helpers.py:21]: Not sure where to throw this comment, but I'd like to make sure we handle user error and add in `https://` if it's not included.

File: looker_blueprints/download_dashboard.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    return args
+
+
+def download_dashboard(sdk, dashboard_id, width, height, file_format):
+    """Download specified dashboard using ID
+    
+    Returns:
+        result: raw binary data of the dashboard
+    """
+    dash_id = dashboard_id
```
Comment (blakeburch) [inline looker_blueprints/download_dashboard.py]: Seems redundant? Just use dashboard_id in the following function?

File: looker_blueprints/download_dashboard.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    task = sdk.create_dashboard_render_task(
+        dash_id,
+        file_format,
+        models.CreateDashboardRenderTask(
+            dashboard_style='tiled',
+            dashboard_filters=None,
+        ),
+        width,
+        height,
+    )
```
Comment (blakeburch) [inline looker_blueprints/download_dashboard.py]: If the dash_id doesn't work, that isn't caught as a specific error. Just prints 404, a bunch of HTML, and returns exit code 1.

File: looker_blueprints/helpers.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    
+    Returns:
+        sdk: Looker 4.0 SDK Class
+    """
+    # initialize looker sdk
+    sdk = looker_sdk.init40(
+    config_settings=MyApiSettings(
+        base_url=base_url,
+        client_id=client_id,
+        client_secret=client_secret
```
Comment (blakeburch) [inline looker_blueprints/helpers.py]: Need error handling for bad credentials. If the any of these don't work, that isn't caught as a specific error. Just prints 404, a bunch of HTML, and returns exit code 1.

File: looker_blueprints/download_look_as_file.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                        required=True)
+    args = parser.parse_args()
+    return args
+
+
+def download_look(look_sdk, look_id, file_format):
+    try:
+        # Options are csv, json, json_detail, txt, html, md, xlsx, sql (raw query), png, jpg
+        response = look_sdk.run_look(
+        look_id=look_id,
```
Comment (blakeburch) [inline looker_blueprints/download_look_as_file.py:41]: Need to handle look_id not existing with a specific error.

File: looker_blueprints/download_look_as_file.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--base-url', dest='base_url', required=True)
+    parser.add_argument('--client-id', dest='client_id', required=True)
+    parser.add_argument('--client-secret', dest='client_secret', required=True)
+    parser.add_argument('--look-id', dest='look_id', required=True)
+    parser.add_argument('--destination-file-name', dest='dest_file_name', required=True)
+    parser.add_argument('--destination-folder-name', dest='dest_folder_name', required=False)
+    parser.add_argument('--file-type', dest='file_type',
+                        choices=['json','txt','csv', 'json_detail', 'md', 'xlsx'
```
Comment (blakeburch) [inline looker_blueprints/download_look_as_file.py]: needs a comma between xlsx and sql

File: looker_blueprints/run_sql_query.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    look_sdk = helpers.get_sdk(base_url, client_id, client_secret)
+    if arg_slug is not None:
+        slug = arg_slug
+    else:
+        try:
+            artifact_subfolder_paths = helpers.artifact_subfolder_paths
+            slug = shipyard.logs.read_pickle_file(artifact_subfolder_paths, 
+                                                        'slug')
+        except Exception as e:
+            print("Error - there was no slug provided and was not found in an upstream vessel. Either enter a slug id for a query already run, or run the 'Create SQL Runner Query' blueprint immediately before this vessel")
```
Comment (blakeburch) [inline looker_blueprints/run_sql_query.py:73]: This doesn't seem to be happening. If I provide an invalid slug, I get a 203 instead.

File: looker_blueprints/run_sql_query.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    return args
+
+
+def run_sql_query_and_download(sdk, slug, file_format):
+    try:
+        response = sdk.run_sql_query(slug=slug, result_format=file_format)
+        print(f"SQL Query {slug} created successfully")
+    except Exception as e:
+        print(f'Error running create query: {e}')
+        sys.exit(ec.EXIT_CODE_LOOK_QUERY_ERROR)
```
Comment (blakeburch) [inline looker_blueprints/run_sql_query.py]: Would like to have a separate error for invalid slug, if possible.

**Review by blakeburch — CHANGES_REQUESTED** [review]
Overall, a few minor adjustments to make, mostly related to error handling. Local CLI functionality looks good overall. 

Once changes are made I'll test in the Shipyard UI.

File: looker_blueprints/download_dashboard.py (hunk 1 above)
Comment (wrp801) [reply looker_blueprints/download_dashboard.py:111]: This is done

File: looker_blueprints/helpers.py (hunk 2 above)
Comment (wrp801) [reply looker_blueprints/helpers.py:21]: This is done

File: looker_blueprints/download_dashboard.py (hunk 3 above)
Comment (wrp801) [reply looker_blueprints/download_dashboard.py]: This is done

File: looker_blueprints/download_dashboard.py (hunk 4 above)
Comment (wrp801) [reply looker_blueprints/download_dashboard.py]: This is done

File: looker_blueprints/helpers.py (hunk 5 above)
Comment (wrp801) [reply looker_blueprints/helpers.py]: This is done

File: looker_blueprints/download_look_as_file.py (hunk 7 above)
Comment (wrp801) [reply looker_blueprints/download_look_as_file.py]: This is done

File: looker_blueprints/download_look_as_file.py (hunk 6 above)
Comment (wrp801) [reply looker_blueprints/download_look_as_file.py:41]: This is done

**Review by blakeburch — APPROVED** [review]
This is good to go.

**Pull request merged**